k-distribution/k-tutorial/1_basic/03_parsing/README.md

```diff
@@ -1 +1,627 @@
+without an `else`. Siknce we want to pick the first parse, we can tell K to
+"avoid" the second parse with the `avoid` attribute. Consider the following
+modified definition (`lesson-03-j.k`):
+
+```k
+module LESSON-03-I
+
+  syntax Exp ::= "true" | "false"
+  syntax Stmt ::= "if" "(" Exp ")" Stmt
+                | "if" "(" Exp ")" Stmt "else" Stmt [avoid]
+                | ";"
+endmodule
+```
+
+Here we have added the `avoid` attribute to the `else` production. As a result,
+when an ambiguity occurs and one or more of the possible parses has that symbol
+at the top of the ambiguous part of the parse, we remove those parses from
+consideration and consider only those remaining. The `prefer` attribute behaves
+similarly, but instead removes all parses which do not have that attribute.
+In both cases, no action is taken if the parse is not ambiguous.
+
+## Tokens
+
+So far we have seen how we can define the grammar of a language. However,
+the grammar is not the only relevant part of parsing a language. Also relevant
+is the lexical syntax of the language. Thus far, we have implicitly been using
+K's automatic lexer generation to generate a token in the scanner for each
+terminal in our grammar. However, sometimes we wish to define more complex
+lexical syntax. For example, consider the case of integers in C: an integer
+consists of a decimal, octal, or hexadecimal number followed by an optional
+suffix indicating the type of the literal.
+
+In theory it would be possible to define this syntax via a grammar, but not
+only would it be cumbersome and tedious, you would also then have to deal with
+an AST generated for the literal which is not convenient to work with.
+
+Instead of doing this, K allows you to define **token** productions, where
+a production consists of a regular expression followed by the `token`
+attribute, and the resulting AST consists of a typed string containing the
+value recognized by the regular expression.
+
+For example, we might define the integer token in C with the following
+production:
+
+```k
+syntax IntConstant ::= r"(([1-9][0-9]*)|(0[0-7]*)|(0[xX][0-9a-fA-F]+))(([uU][lL]?)|([uU]((ll)|(LL)))|([lL][uU]?)|(((ll)|(LL))[uU]?))?" [token]
+```
+
+As you can see, the main difference with this production is the `r` preceding
+the terminal and the `token` attribute. The `r` indicates that what follows
+is a regular expression. It is also possible to define tokens that do not
+use regular expressions. This can be useful when you wish to declare particular
+identifiers for use in your semantics later. For example:
+
+```k
+syntax Id ::= "main" [token]
+```
+
+Here we declare that `main` is a token of sort `Id`. Instead of being parsed
+as a symbol, it gets parsed as a token instead, generating a typed string
+in the AST. This is useful because the parser generally does not treat the
+`main` function in C specially; only the semantics treats it specially.
```

Comment:
```suggestion
Here we declare that `main` is a token of sort `Id`. Instead of being parsed
as a symbol, it gets parsed as a token, generating a typed string
in the AST. This is useful because the parser generally does not treat the
`main` function in C specially; only the semantics treats it specially.
```